### PR TITLE
Implement DDQN

### PR DESCRIPTION
**Pull request by gonzaiva**
An optimal policy is derived by selecting the highest-valued action in
each state of the environment, namely, using Q-learning. Although a
policy is not required since it's obtained from the Q-value function,
Q-learning is limited to act in discrete action spaces.
DDQN has two important ingredients:                                             
    - A target network. Although the online network makes the
    predictions, the target network helps in the training of the online
    network. The parameters of the target networks are copied every
    tau steps from the online network and kept fixed on all other steps.
    This reduces the overestimations produced in Q-learning.                                     
    - Experience replay. The observed transitions are stored for some
    time and sampled uniformly from a memory bank to update the
    network.

### CODE REVIEW

**Comment by ryanjulian** [comment]
Please post benchmark results comparing to OpenAI baselines. You can base it Chang's script. See my comments on her script about setting random seeds in garage and TF

File: garage/algos/ddqn.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+            reward_episode += reward
+
+    def _copy_to_target(self):
+        """Copy weights from online network to target network."""
+        ll.set_all_param_values(self._target_nn,
+                                ll.get_all_param_values(self._online_nn))
+
+
+class MemoryBank:
+    """The memory bank is used for experience replay in DDQN."""
```
Comment (zhanpenghe) [inline garage/algos/ddqn.py]: I think this can be moved to garage.replay_buffer

Comment (zhanpenghe) [reply garage/algos/ddqn.py]: Also there is a replay_buffer in `garage.algos.ddpg` implementation already so if you can just move it out as a module and make use fo it, that would be great.

File: garage/algos/ddqn.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+            Number of transitions to put in the memory batch.
+
+        """
+        random_indices = np.random.randint(0, self._num_transitions, n)
+        obs = [self._obs_arr[index] for index in random_indices]
+        obs_next = [self._obs_next_arr[index] for index in random_indices]
+        actions = [self._action_arr[index] for index in random_indices]
+        rewards = [self._reward_arr[index] for index in random_indices]
+        not_done = [self._not_done_arr[index] for index in random_indices]
+        return MemoryBatch(obs, obs_next, actions, rewards, not_done)
```
Comment (zhanpenghe) [inline garage/algos/ddqn.py]: I think a dictionary would be enough for this purpose.

File: garage/algos/ddqn.py [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+        self._action = tensor.ivector("Action")
+        self._reward = tensor.vector("Reward")
+        self._done = tensor.vector("Done")
+        # Build the networks
+        self._target_nn = self._create_network()
+        self._online_nn = self._create_network()
+        self._build_graph()
+
+    def _create_network(self):
+        """Create a neural network.
```
Comment (zhanpenghe) [inline garage/algos/ddqn.py]: I think this should be moved to something like q_function. Please remember to create some reusable components.

**Comment by ryanjulian** [comment]
Cleaning up. Please reopen when it's rebased and ready for review.

**Pull request closed without merging**